### PR TITLE
librbd: snapshot object maps can go inconsistent during copyup

### PR DESCRIPTION
**Pull request by idryomov**

### CODE REVIEW

**Comment by idryomov** [comment]
I see some rbd test failures, looking...

File: src/librbd/io/CopyupRequest.cc [hunk 1]
```diff
@@ -233,6 +234,7 @@ void CopyupRequest<I>::handle_deep_copy(int r) {
 
   m_image_ctx->snap_lock.get_read();
   m_lock.Lock();
+  m_copyup_is_zero = m_copyup_data.is_zero();
```
Comment (dillaman) [inline src/librbd/io/CopyupRequest.cc]: Nit: the ```m_copyup_data``` will be empty here after a deep-copy, so ```m_copyup_is_zero``` can just be hard-coded to true.

File: src/librbd/io/CopyupRequest.cc [hunk 2]
```diff
@@ -291,7 +291,7 @@ void CopyupRequest<I>::update_object_maps() {
   auto cct = m_image_ctx->cct;
   ldout(cct, 20) << "oid=" << m_oid << dendl;
 
-  if (!m_image_ctx->snaps.empty()) {
+  if (!m_image_ctx->snaps.empty() && !m_copyup_is_zero) {
```
Comment (dillaman) [inline src/librbd/io/CopyupRequest.cc]: ... again, the deep-copy is the odd case where we *might* need to update the object map snapshots if deep-copy actually copied data (i.e. the object exists).

File: src/librbd/io/CopyupRequest.cc [hunk 3]
```diff
@@ -269,6 +279,14 @@ void CopyupRequest<I>::handle_deep_copy(int r) {
     return;
   }
 
+  // For deep-copy, copyup() will never affect snapshots.  However,
+  // this state machine is responsible for updating object maps for
+  // snapshots that have been created on destination image after
+  // migration stared.
```
Comment (dillaman) [inline src/librbd/io/CopyupRequest.cc]: Nit: ```started```

**Comment by idryomov** [comment]
> lgtm -- there might be one additional bug w.r.t. deep-copy sets the first post-migration snapshot object map to `EXISTS` instead of `EXISTS_CLEAN`

Yes, I think you are right.  If there are pre-migration snapshots, then it should be set to `EXISTS_CLEAN`, otherwise to `EXISTS`.

Would inferring this info from `compute_deep_copy_snap_ids()` work?  I.e. if any snapshot is skipped because its object map is already updated (`deep_copy` set), the first snapshot object map state should be `EXISTS_CLEAN`, otherwise `EXISTS`.

**Comment by dillaman** [comment]
> Would inferring this info from `compute_deep_copy_snap_ids()` work?

Yeah, that sounds good to me.

**Comment by idryomov** [comment]
> > Would inferring this info from `compute_deep_copy_snap_ids()` work?
> 
> Yeah, that sounds good to me.

Updated.

**Pull request merged**